src/levanter/eval_harness.py

```diff
@@ -0,0 +1,564 @@
+            targets = hax.roll(example.tokens, -1, axis=Pos)
+            freebie = hax.logical_not(example.loss_mask * not_last_loss_mask)
+            correct = hax.all(hax.equal(pred_targets, targets) + freebie, axis=Pos)
+
+            return -loss, correct
+
+        # no sharded outputs
+        self._jit_loglikelihood = hax.named_jit(
+            _eval_loglikelihood, axis_resources=axis_resources, out_axis_resources={}
+        )
+
+    def loglikelihood(self, requests: list[Instance]) -> list[tuple[float, bool]]:
+        """
+        Compute log-likelihood of generating a continuation from a context.
+        Downstream tasks should attempt to use loglikelihood instead of other
+        LM calls whenever possible.
+        """
+        # pad requests to be a multiple of the batch size
+        initial_length = len(requests)
+        dummy_instance = dataclasses.replace(requests[0], arguments=("hello", " there"), idx=len(requests))
+        requests = requests + [dummy_instance] * (self.EvalBatch.size - len(requests) % self.EvalBatch.size)
+        assert len(requests) % self.EvalBatch.size == 0
+        dataset = EvalDataset(self.EvalPos, self.tokenizer, requests)
+
+        mesh = haliax.partitioning._get_mesh()
+
+        loader = DataLoader(
+            self.EvalBatch, dataset, max_buffered_batches=1024, mesh=mesh, axis_resources=self.axis_resources
+        )
+
+        result: list[tuple[float, bool]] = []
+        for batch in tqdm(loader, desc="Loglikelihood", unit="ba"):
+            out_lls, out_correct = self._jit_loglikelihood(self.model, batch)
+            result.extend((ll.item(), correct.item()) for ll, correct in zip(out_lls.array, out_correct.array))
+
+        assert len(result) >= initial_length
+        # skip padding
+        result = result[:initial_length]
+
+        return result
+
+    def loglikelihood_rolling(self, requests) -> List[Tuple[float]]:
+        raise NotImplementedError()
+
+    def generate_until(self, requests) -> List[str]:
+        raise NotImplementedError()
+
+
+@functools.partial(jax.jit, static_argnums=(0, 3))
+def _jit_create_example(Pos, tokens, prompt_len, pad_token_id):
+    tokens = hax.named(tokens, Pos)
+    return LmExample.from_prompt_and_completion(Pos, tokens, prompt_len, ignore_id=pad_token_id)
+
+
+@dataclass(frozen=True)
+class TaskConfig:
+    """
+    This is a dataclass that represents the configuration for a task in the LM Eval Harness. It is used to specify
+    the configuration for a task in the LM Eval Harness, and is used to generate the task dictionary that the LM Eval
+    Harness expects.
+
+    nb that LM Eval Harness has its own TaskConfig, but its defaults are not the same as just passing in
+    a dict, and we want the behavior of passing in a dict.
+
+    See Also:
+        [LM Eval Harness TaskConfig](https://github.com/EleutherAI/lm-evaluation-harness/blob/0ef7548d7c3f01108e7c12900a5e5eb4b4a668f7/lm_eval/api/task.py#L55)
+    """
+
+    task: str
```

Comment:
Can we add docs for each field?

Reply:
i mean, they don't even have docs for a lot of these.